### ISSUE TITLE
vision: diffusion training on MPS needs a group_norm backward

### ISSUE THREAD

**Issue by bkc39** [bkc39/rktorch#139]
`examples/racket/08-diffusion.rkt` (#138) keeps its training run on the CPU when the accelerator is MPS: libtorch 2.9's MPS backend has no kernel for `aten::native_group_norm_backward`, so the first `backward!` through a `GroupNorm` raises (the macOS CI job on #138 showed it before the fallback). The forward of every layer in the UNet, the CIFAR-10 dataset on MPS, and the loader all work there; only training of a model that contains `GroupNorm` is affected.

Options, none taken yet:

- Wait for a libtorch that ships the kernel (the MPS backend gains ops release by release) and drop the fallback then.
- A `group-norm` implementation from primitives that do have MPS backward (reshape to `[N G -1]`, mean and variance by reduction, the affine at the end), selected when the input lives on MPS. Costs a second code path and a parity twin of its own.
- `PYTORCH_ENABLE_MPS_FALLBACK=1` in the process environment before libtorch loads, which makes ATen run the missing op on the CPU; it must be set before the shim is loaded, so it belongs in the nix shell or the launcher, not in Racket.

Refs: #84, #138.

**Comment by bkc39** [comment]
Taken in #138 as a carve-out, the same way ctc-loss handles its MPS gap: the GroupNorm layer runs the normalisation on the CPU through differentiable to-device copies when its input is on MPS, so the UNet trains on the GPU there with the norms round-tripped. This issue now tracks removing that round trip once libtorch ships native_group_norm_backward for MPS.

**Comment by bkc39** [comment]
Upstream status: the Metal kernel exists. pytorch/pytorch#183830 added `native_group_norm` for MPS (merged 2026-05-22) and pytorch/pytorch#184437 added `native_group_norm_backward` (merged 2026-05-27); the dispatch table lists MPS for both from v2.13.0 (2026-07-08) on, and v2.14.0 (2026-09-02) is current. Our libtorch-bin is 2.9.0 and the parity Python is 2.12, neither of which has it. The carve-out in the GroupNorm layer can go, and the vendored schema be refreshed, when the nixpkgs pin moves the libtorch to 2.13 or later.

**Comment by bkc39** [comment]
Two notes from the MPS run of #147, one a correction to this issue's text and one a measurement of what the workaround costs.

**The fallback is narrower than the issue says.** The body reads "keeps its training run on the CPU when the accelerator is MPS". The shipped implementation does not do that. `torch/nn/group-norm.rkt` detours only the group-norm itself:

```racket
(if (eq? (device-type device) 'mps)
    (to-device (group-norm (to-device x 'cpu) num-groups
                           #:weight (to-device weight 'cpu)
                           #:bias (to-device bias 'cpu)
                           ...)
               device)
    ...)
```

`to-device` is differentiable both ways, so the gradient returns across the hop. Everything else in the UNet, the convolutions, the attention, the optimizer and the loss, runs on MPS. The example's own prose says as much; only this issue's summary implies a whole-run fallback. Worth correcting, because it changes what an MPS run of that example proves: it is genuine MPS training with one op detoured, not CPU training with the data parked on the device.

**What the detour costs.** The UNet builds a `GroupNorm` in each `ResBlock` twice, once per `AttentionBlock`, and once at the output, so a forward crosses the device boundary a few dozen times, and the backward crosses it again. The #147 run on an M2 Pro at base 128 and batch 64 gave a median of 2.16 s per step early and 2.66 s later, with MPS allocated at 2.4 GB throughout, which is consistent with the round-trips dominating rather than with any single kernel being slow.

That makes the libtorch bump in #140 worth more than a compatibility update. Removing the detour should take a large bite out of per-step time on Apple silicon, not merely tidy the code. It is the one change that would make MPS training of this model competitive rather than merely possible.

I have no Apple hardware here, so I cannot measure the difference, and I am not claiming a figure for it. Someone with a Mac could bound it cheaply by timing the same example with the detour forced off against a model with the norms removed.

Refs #140, #147.
